File: c++/leetcode/817. Linked List Components.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
  int numComponents(ListNode *head, vector<int> &G)
  {
    unordered_set<int> seen{G.begin(), G.end()};
    int connections{};

    while (head)
    {
      if (seen.count(head->val) && (!head->next || !seen.count(head->next->val)))
        ++connections;

      head = head->next;
    }

    return connections;
  }
};

File: javascript/leetcode/817. Linked List Components.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number[]} G
 * @return {number}
 */
var numComponents = function (head, G) {
  const seen = new Set(G);
  let connections = 0;

  while (head) {
    if (seen.has(head.val) && (!head.next || !seen.has(head.next.val)))
      ++connections;

    head = head.next;
  }

  return connections;
};
